### PR TITLE
Pennylane is compatible with numpy 2.0

### PR DESCRIPTION
**Pull request by EmilianoG-byte**
**Context:**

We want to make Pennylane compatible with Numpy 2.0. After several discussions, we decided to test NumPy 2.0 on the CI by default in every PR (testing both Python versions would have been to slow). Some jobs still downgrade automatically to Numpy 1.x, since some interfaces (such as Tensorflow) still do not support NumPy 2.0.

**Description of the Change:**

We can distinguish the changes into 3 main categories:

*Changes to workflows*

- None in the final version

*Changes to requirements and setup files*

- Unpin the Numpy version in `setup.py` (now we also allow Numpy 2.0).
- Update `requirements-ci.txt` to include Scipy 1.13 (this adds support for Numpy 2.0).

*Changes to the source code*
- Change `np.NaN` to `np.nan`.
- Use legacy printing representation in tests, contrary to the new numpy representation of scalars, e.g. np.float64(3.0) rather than just 3.0.
- Update probabilities warning to be case insensitive and check for a partial match, since this warning was changed in Numpy 2.0.
- Check the datatype of np.exp from the Global phase only for Numpy 1.x, since this gets promoted to complex128 in Numpy 2.x. https://numpy.org/neps/nep-0050-scalar-promotion.html#schema-of-the-new-proposed-promotion-rules.

**Benefits:**

Make Pennylane compatible with Numpy 2.0.

**Possible Drawbacks:**

- We need to create a separate workflow to keep testing PennyLane with NumPy 1.x, since we still want to maintain compatibility with previous NumPy versions. This will be done in a separate PR.

- We are not testing Numpy 2.x for the interfaces that implicitly require Numpy 1.x. These currently seem to be `tensorflow` and `openfermionpyscf` (notice that `tensorflow` is required in some code sections like qcut).  In particular, `openfermionpyscf` causes an error: 
```
AttributeError: np.string_ was removed in the NumPy 2.0 release. Use np.bytes_ instead.
``` 
in the qchem tests. The attribute `np.string_` is not used in the PL source code, so it is a problem with the package itself.

[sc-61399]
[sc-66548]

### CODE REVIEW

File: .github/workflows/install_deps/action.yml
Comment (astralcai) [inline .github/workflows/install_deps/action.yml:20]: Why do we add this to the workflows as opposed to pinning a `numpy` version in `requirements.txt`?

File: pennylane/numpy/random.py
Comment (astralcai) [inline pennylane/numpy/random.py:33]: Is this meant to be removed?

**Review by astralcai — COMMENTED** [review]
Is this PR only fixing tests?

File: tests/data/attributes/operator/test_operator.py
Comment (astralcai) [inline tests/data/attributes/operator/test_operator.py:222]: Maybe instead of the entire test function, only put the affected lines in the context manager, so
```
with np.printoptions(legacy="1.21"):
    assert repr(op_out) == repr(op_in)
```
and keep the rest of the test case untouched.

File: pennylane/numpy/random.py
Comment (EmilianoG-byte) [reply pennylane/numpy/random.py:33]: This is something I was still waiting to get @mlxd comment on since I got it from his original PR and I am not sure if it's needed.

File: .github/workflows/install_deps/action.yml
Comment (EmilianoG-byte) [reply .github/workflows/install_deps/action.yml:20]: If I understand this correctly, this is needed because we are expecting to be able  to test Pennylane both with Numpy 1.x and 2.x, so this action is in preparation to be able to install one or the other version.

Comment (mlxd) [reply .github/workflows/install_deps/action.yml:20]: Exactly that --- we need to run tests on both the 1.x and 2.x series.

File: .github/workflows/install_deps/action.yml
Comment (mlxd) [inline .github/workflows/install_deps/action.yml:28]: This will need to be 0.4.28, right?

File: .github/workflows/install_deps/action.yml
Comment (mlxd) [inline .github/workflows/install_deps/action.yml:36]: No issues with TF failing, or are we not testing that combination?

File: .github/workflows/install_deps/action.yml
Comment (mlxd) [inline .github/workflows/install_deps/action.yml:78]: Does this validate the `default` argument when not specified?

File: .github/workflows/install_deps/action.yml
Comment (mlxd) [inline .github/workflows/install_deps/action.yml:80]: What does this string evaluate to? Is it `=2.0.0`  when set?

File: .github/workflows/interface-unit-tests.yml
Comment (mlxd) [inline .github/workflows/interface-unit-tests.yml:148]: Do we want to run `[false, true]` here, or is that too expensive?
Also, shouldn't this be used in every interface test, and not just torch?

File: pennylane/numpy/random.py
Comment (mlxd) [reply pennylane/numpy/random.py:33]: This can be removed --- it was used temporarily as the numpy wheels were not semver compliant. We can add this in without problem, but safe to assume wheels should comply.

File: requirements-ci.txt
Comment (mlxd) [inline requirements-ci.txt:1]: We likely shouldn't pin the version here.

File: requirements-ci.txt
Comment (mlxd) [inline requirements-ci.txt:2]: Can this also be unpinned, or is it required?

File: requirements-ci.txt
Comment (mlxd) [inline requirements-ci.txt:7]: No issues when running the latest autograd against 1.x python?

File: tests/data/attributes/operator/test_operator.py
Comment (mlxd) [inline tests/data/attributes/operator/test_operator.py:177]: Do we want 1.21 or 1.25 here?
I'd also argue this needs to have a ticket created (to be handled another time) for updating the datasets to the new syntax @albi3ro @Alex-Preciado

File: tests/devices/test_default_qubit_legacy.py
Comment (mlxd) [inline tests/devices/test_default_qubit_legacy.py:657]: I wonder if it would be better to wrap these in a contextmanager + fixture? Though, fine by me to use the given style

**Review by mlxd — COMMENTED** [review]
Thanks @EmilianoG-byte for all of the work 
A few questions from my side:

File: .github/workflows/install_deps/action.yml
Comment (astralcai) [reply .github/workflows/install_deps/action.yml:20]: I believe `requirements.txt` still pins `numpy` to `1.23`. Should we unpin that?

File: tests/data/attributes/operator/test_operator.py
Comment (EmilianoG-byte) [reply tests/data/attributes/operator/test_operator.py:177]: The problem with using 1.25 is what I mentioned in the slack channel, because in some action script gets downgraded to scipy < 1.13, which then downgrades numpy to < 2.0, such that 1.25 is not available as an option anymore.

File: pennylane/numpy/random.py
Comment (EmilianoG-byte) [reply pennylane/numpy/random.py:33]: Sounds good, I will erase it then.

File: .github/workflows/install_deps/action.yml
Comment (EmilianoG-byte) [reply .github/workflows/install_deps/action.yml:20]: We had a conversation about this in a thread with @timmysilv, quoting:
> requirements.txt can serve as a minimum version for dependencies, whereas requirements-ci.txt allows CI to test with the maximum version

So maybe we can change `numpy~=1.23` to `numpy>=1.23` ?

File: .github/workflows/install_deps/action.yml
Comment (EmilianoG-byte) [reply .github/workflows/install_deps/action.yml:28]: I believe this might be a different story: https://app.shortcut.com/xanaduai/story/61389/ensure-pennylane-is-compatible-with-jax-0-4-26

File: .github/workflows/install_deps/action.yml
Comment (mudit2812) [reply .github/workflows/install_deps/action.yml:80]: If set, it would be `numpy~=2.0.0`

File: requirements-ci.txt
Comment (EmilianoG-byte) [reply requirements-ci.txt:2]: For Scipy? I also believe it shouldn't have been pinned from the beginning according to the dependency management guidelines: https://www.notion.so/xanaduai/Dependency-Management-0756a846d19d42e490fa233afc08ca2b?pvs=4

File: requirements-ci.txt
Comment (EmilianoG-byte) [reply requirements-ci.txt:7]: Against Numpy 1.x? I don't think we have a separate workflow to test the newest autograd with the older number versions.

File: requirements-ci.txt
Comment (EmilianoG-byte) [reply requirements-ci.txt:1]: I thought we wanted to keep it at some 2.0.x version to have it compatible with Autograd

File: tests/data/attributes/test_dict.py
Comment (mudit2812) [inline tests/data/attributes/test_dict.py]: If `dset_dict` is set inside the context manager, we shouldn't have to put these assertions in the context manager as well.

Comment (mudit2812) [reply tests/data/attributes/test_dict.py]: Same comment for other assertions as well.

File: tests/devices/test_default_qubit_legacy.py
Comment (mudit2812) [reply tests/devices/test_default_qubit_legacy.py:657]: That's a cool idea. I think that for the limited number of tests that are impacted by this, it's not necessarily needed, but it's probably a great way to future proof the test suite. I convinced myself that I like this change as I was typing out this comment 😄

**Review by mudit2812 — COMMENTED** [review]
Looks good, I left a few comments and suggestions.

File: .github/workflows/install_deps/action.yml
Comment (josh146) [reply .github/workflows/install_deps/action.yml:28]: Yep we should update this to 0.4.28 once the release is made (as Catalyst has been updated to work with this JAX version)

File: .gitignore
Comment (PietropaoloFrisoni) [reply .gitignore:31]: I'll push it in the next commit

File: .github/workflows/install_deps/action.yml
Comment (PietropaoloFrisoni) [reply .github/workflows/install_deps/action.yml:20]: @mlxd  Please let me know if you have issues with keeping `numpy` pinned to 1.23 in `requirements.txt`, otherwise we can mark this conversation as resolved

File: .github/workflows/install_deps/action.yml
Comment (PietropaoloFrisoni) [reply .github/workflows/install_deps/action.yml:28]: Looks like updating JAX to 0.4.28 broke a lot of tests (thanks @mudit2812 for the [investigation](https://xanaduhq.slack.com/archives/C0271MF2M50/p1725474140063309)). At this point, we should probably open a second PR once this is merged, hopefully after [this PR](https://github.com/PennyLaneAI/pennylane/pull/6215) from @albi3ro is merged.

File: .github/workflows/install_deps/action.yml
Comment (PietropaoloFrisoni) [reply .github/workflows/install_deps/action.yml:84]: Yep, `~=2.0.0` if `inputs.numpy_2_version` is `2.0.0` (default value), otherwise an empty string if `inputs.numpy_2_version` is also an empty string

File: .github/workflows/install_deps/action.yml
Comment (PietropaoloFrisoni) [reply .github/workflows/install_deps/action.yml:82]: @mlxd I am not sure I understand the question. Are you asking what happens to line 84 (that is, to `NUMPY2_VERSION)` if `numpy_2_version` is not specified or `inputs.install_numpy_2 ` is false?

File: tests/devices/test_default_qubit_legacy.py
Comment (PietropaoloFrisoni) [reply tests/devices/test_default_qubit_legacy.py:657]: > it's probably a great way to future proof the test suite

I think it might be something to add to the [new epic](https://app.shortcut.com/xanaduai/epic/72982?group_by=none&ct_workflow=all&cf_workflow=500000005) for CI improvements or the [epic](https://app.shortcut.com/xanaduai/epic/66006?group_by=none&vc_group_by=day&ct_workflow=all&cf_workflow=500000005) for improving robustness of tests + CI. I'll resolve the conversation since It seems out of the scope of the original PR.

Comment (mudit2812) [reply tests/devices/test_default_qubit_legacy.py:656]: The 2nd epic you linked is probably more appropriate. The first one is more targeted towards updating the workflows, not the tests.

File: .github/workflows/interface-unit-tests.yml
Comment (mudit2812) [inline .github/workflows/interface-unit-tests.yml:1]: So basically every single test will always run with both numpy 2.0 and 1.x. That's quite a lot more load on CI.

File: pennylane/numpy/random.py
Comment (mudit2812) [inline pennylane/numpy/random.py:19]: Why was this added?

File: .github/workflows/interface-unit-tests.yml
Comment (mudit2812) [reply .github/workflows/interface-unit-tests.yml:1]: Missed the updated PR description. Feel free to disregard this comment.

**Review by mudit2812 — APPROVED** [review]
Looks good. No blockers from my side. I'll approve, but this PR should probably require 3 approvals to be merged.

File: pennylane/numpy/random.py
Comment (PietropaoloFrisoni) [reply pennylane/numpy/random.py:19]: There was a conflict between isort and CodeFactor

**Comment by PietropaoloFrisoni** [comment]
> but this PR should probably require 3 approvals to be merged.

@mudit2812 I agree. And if we are 100% sure to test everything twice, maybe it's best to announce this on the pennylane channel before we merge (an action that in any case requires to force-merge)

**Comment by mlxd** [comment]
Hey @PietropaoloFrisoni 
Thanks for the additional work here. I noticed that even though the CI appears green for many of the `numpy2.0 true` runs, we are still using numpy 1.x in the runs themselves (e.g. check the logs for many of them, which reports the installed numpy version). I think if we don't have numpy 2.0 support included for a given interface, we can disable the `true` runs for such interfaces, since those aren't correctly reporting. Sound reasonable?

**Comment by PietropaoloFrisoni** [comment]
Update: now, Codecov project complains since it detects a -0.07% coverage compared to the commit associated with [this PR](https://github.com/PennyLaneAI/pennylane/pull/6209). As specified in the PR itself, _Code cov failures are in default.qubit.legacy and QubitDevice, so can be ignored._

**Review by PietropaoloFrisoni — APPROVED** [review]

**Pull request merged**
